### ISSUE TITLE
How do I ignore specific classes in an image directory tree

### ISSUE THREAD

**Issue by tszumowski** [PyTorchLightning/lightning-flash#763]
How do I ignore specific classes in a directory tree? 
Suppose I have a directory tree with 500 classes as sub directories: `class_001` ... `class_500`. But I want to build a classifier on just two classes: `class_001` and `class_003`.

Suppose I am loading datasets via fifty one like:

```python
labels_fo = ["class_001", "class_003"]
train_dataset = fo.Dataset.from_dir(
    dataset_dir=train_dir,
    dataset_type=fo.types.ImageClassificationDirectoryTree,
    name="training_dataset"
)
# Filter based on labels of interest.
train_dataset = train_dataset.match(F("ground_truth.label").is_in(labels_fo))
train_dataset.save()
```

And then I create a Data Module with:

```python
datamodule = ImageClassificationData.from_fiftyone(
    train_dataset=train_dataset,
    val_split=0.2,
    batch_size=64
)
```

In this example, the FiftyOne dataset is filtered down to two classes: `class_001` and `class_003`. But the `datamodule.num_classes = 3`. This is because [find_classes](https://github.com/PyTorchLightning/lightning-flash/blob/3b88e74e54cf8269e41859e3d218973d9e1e9979/flash/core/data/data_source.py#L431-L443) method runs a `scan_dir` on the folder, without considering the labels in the dataset.

**Comment by dlangerm** [comment]
@tszumowski Would it be too much to just create a separate directory (even just symlink one and avoid a copy) with just your data? It seems like an unintuitive use case to have a directory tree class that doesn't just reflect back the underlying directory.

**Comment by tszumowski** [comment]
Hi @dlangerm thank you for the reply. A symlink directory is what I ended up doing last week.
In my use case, I have a directory tree with hundreds of classes and lots of images. But I am creating classifiers for sub-sets or different groupings of those classes. 

The symlink method works, but ends up with a bunch of permutations of symlinked directories on disk. I figured I'd check to see if there was a class or label filter.

Where I ended up gettog is a fiftyone dataset offers the ability to filter a dataset by the label (class). So I was able to load that full dataset using `from_dir`. Then I was able to filter down using FiftyOne's filter dataset using FiftyOne's library. But that filter didn't stay consistent with the PyTorch Lightning Flash representation when I trained the model. i.e the model said 
 it had hundreds of classes, even though I filtered down to two in the FiftyOne object.

**Comment by dlangerm** [comment]
Hm. @tszumowski I have to agree that sounds like a bug to me. Any ideas on a pull request that would fix it to remain consistent?

**Comment by tszumowski** [comment]
> Hm. @tszumowski I have to agree that sounds like a bug to me. Any ideas on a pull request that would fix it to remain consistent?

I'm still fairly new to the Lightning-Flash and FiftyOne libraries. So I can't point to internals. But the `ImageClassificationData.from_fiftyone(` can either:
a) Filter down to the defined labels in the FiftyOne dataset internally, or
b) Request a user provide an optional `labels=` input argument to define what to filter down to.

Option (a) keeps it internally consistent between the two libraries. I'm just not sure how that aligns with `from_fiftyone` and all the other dataset types that Lightning-Flash must concurrently support.

**Comment by dlangerm** [comment]
I'm thinking option (a) makes a lot of sense and should be straightforward to implement.

**Comment by brimoor** [comment]
@tszumowski Actually you just need to add `train_dataset.default_classes = None` before creating your Lightning data module:

```py
import fiftyone as fo
import fiftyone.zoo as foz
from fiftyone import ViewField as F

from flash.image import ImageClassificationData

dataset = foz.load_zoo_dataset("cifar10", split="test", max_samples=51, shuffle=True, seed=51)

dataset.export(
    export_dir="/tmp/dataset",
    dataset_type=fo.types.ImageClassificationDirectoryTree,
)

train_dataset = fo.Dataset.from_dir(
    dataset_dir="/tmp/dataset",
    dataset_type=fo.types.ImageClassificationDirectoryTree,
)

classes = ["airplane", "automobile"]
train_dataset.match(F("ground_truth.label").is_in(classes)).save()

datamodule = ImageClassificationData.from_fiftyone(train_dataset=train_dataset)
print(datamodule.num_classes)  # 10

# Clear default classes
# Now classes will be computed from actual dataset contents
train_dataset.default_classes = None

datamodule = ImageClassificationData.from_fiftyone(train_dataset=train_dataset)
print(datamodule.num_classes)  # 2
```

FiftyOne datasets [can store class lists](https://voxel51.com/docs/fiftyone/user_guide/using_datasets.html#storing-class-lists) (used for various things like evaluation), and when you import a FiftyOne dataset from a classification directory tree, the observed classes are in fact populated. Lightning Flash [uses those classes](https://github.com/PyTorchLightning/lightning-flash/blob/d190c7675ccc2198fe954828a50a29c9a96006f2/flash/core/data/data_source.py#L701-L711), if present, when constructing the data module.

In your case, since you've filtered your FiftyOne dataset, erasing the stored classes list will cause the observed classes to be used instead (or equivalently you could update `default_classes` on your dataset).

**Comment by dlangerm** [comment]
Hm. Still seems like a bug to me. If the data is coming from the fifty-one dataset, then intuitively the number of classes should be sources from it as well. Un-setting a default value to change where that information is coming from feels like a hack and not a fix.
